diff --git a/Aeneas/Std/Core/Ptr.lean b/Aeneas/Std/Core/Ptr.lean
index 7e824e1d..e3e952f7 100644
--- a/Aeneas/Std/Core/Ptr.lean
+++ b/Aeneas/Std/Core/Ptr.lean
@@ -73,7 +73,7 @@ inductive core.ptr.alignment.AlignmentEnum where
 | _Align1Shl62 : core.ptr.alignment.AlignmentEnum
 | _Align1Shl63 : core.ptr.alignment.AlignmentEnum
 
-@[reducible, rust_type "core::ptr::alignment::Alignment"]
+@[reducible, expose, rust_type "core::ptr::alignment::Alignment"]
 def core.ptr.alignment.Alignment :=
 core.ptr.alignment.AlignmentEnum
 
diff --git a/Aeneas/Data/Array.lean b/Aeneas/Data/Array.lean
index 63938a02..0fdb0077 100644
--- a/Aeneas/Data/Array.lean
+++ b/Aeneas/Data/Array.lean
@@ -110,13 +110,7 @@ theorem getElem_set! {α : Type u}
   simp only [set!_eq_setIfInBounds, ← getElem_toList, toList_setIfInBounds]
   simp_lists
 
-@[simp, simp_lists_safe]
-theorem getElem!_set!_ne {α : Type u}
-  [Inhabited α] {i j : ℕ} {x : α} {xs : Array α}
-  (h : i ≠ j) :
-  (xs.set! i x)[j]! = xs[j]! := by
-  simp only [set!_eq_setIfInBounds, ← getElem!_toList, toList_setIfInBounds]
-  simp_lists
+attribute [simp, simp_lists_safe] getElem!_set!_ne
 
 @[simp, simp_lists_safe]
 theorem getElem_set!_ne {α : Type u}
